Event/event.js

```diff
@@ -1,17 +1,81 @@
+var BusyEvents = [];
+const NEXT_WEEK=7*24*60*60*1000; //same time next week in mill sec
+const NEXT_SLOT=30*60*1000; //next 30 min in mill sec
 
-var Event = function(opening, recurring, startDate, endDate){
+var Event = function (opening, recurring, startDate, endDate) {
     this.opening = opening;
     this.recurring = recurring;
     this.startDate = startDate;
     this.endDate = endDate;
-
     eventList.push(this);
 };
 
-Event.prototype.availabilities = function(fromDate, toDate){
-    return //Something awesome;
+Event.prototype.availabilities = function (fromDate, toDate) {
+    eventList.forEach(function (element) {
+        if (element.opening === true) {
+            //  get free slot (start & end) in mill sec
+            start_open = Date.parse(element.startDate);
+            end_open = Date.parse(element.endDate);
+            AddFreeEvents(element, Date.parse(fromDate), Date.parse(toDate), start_open, end_open);
+        };
+    });
+    eventList.forEach(function (element) {
+        if (element.opening === false) {
+            //  get busy windows (start & end) in mill sec
+            start_busy = Date.parse(element.startDate);
+            end_busy = Date.parse(element.endDate);
+            AddBusyEvents(element, Date.parse(fromDate), Date.parse(toDate), start_busy, end_busy);
+        };
+    });
+    //  available windows = free windows - busy windows
+    let availabilities = FreeEvents.filter(element => !BusyEvents.includes(element));
+
+    //print availabilities
+    if (availabilities.length == 0) {
+        console.log('I\'m not available at all.');
+    } else {
+        availabilities.forEach(function (slot) {
+            console.log('I\'m available the ' + slot);
+        });
+        console.log('I\'m not available any other time !');
+    }
 };
 
-module.exports = Event;
+function AddFreeEvents(element, fromDate, toDate, start_open, end_open) {
+    start_date=Math.max(fromDate,start_open);
+    end_date=Math.min(toDate,end_open);
+    next_slot = start_date;
+    while (next_slot < end_date){
+        FreeEvents.push(timeToString(next_slot));
+        next_slot += NEXT_SLOT; // check next 30 min slot
+    };
+    //  recursive call for recurring available events
+    if (element.recurring === true && start_open < toDate) {
+        AddFreeEvents(element, fromDate, toDate, start_open + NEXT_WEEK, end_open + NEXT_WEEK);
+    }
+}
+function AddBusyEvents(element, fromDate, toDate, start_busy, end_busy) {
+    start_date=Math.max(fromDate,start_busy);
+    end_date=Math.min(toDate,end_busy);
+    next_slot = start_date;
+    while (next_slot < end_date) {
+        BusyEvents.push(timeToString(next_slot));
+        next_slot += NEXT_SLOT; // check next 30 min slot
+    };
+
+    //  recursive call for recurring busy events
+    if (element.recurring === true && start_busy < toDate) {
+        AddBusyEvents(element, fromDate, toDate, start_busy + NEXT_WEEK, end_busy + NEXT_WEEK);
+    }
+}
 
```

Comment:
These 2 methods are really similar, I think we could have refactored them

Reply:
True. After rethinking, this could have been improved. For example, by passing the array to be filled as a last parameter depending on the opening state (true or false).
The function could be AddEvents(element, fromDate, toDate, startBusy, endBusy, arrayToBeFilled)